netutils/os_version_parser.py

```diff
@@ -0,0 +1,90 @@
+
+    Examples:
+        >>> parsed_version = juniper_junos_version_parser("12.3R4")
+    """
+    # Use regex to group the main, minor, type and build into useable pieces
+    # re_main_minor_type_build = re.search(r"^(\d+)\.(\d+)([xXrRsS])?(\d+)?", split_version[0])
+    re_main_minor_type_build: re.Pattern[str] = re.compile(
+        r"""
+        ^
+        (?P<main>\d+)           # main train
+        \.                      # dot separator
+        (?P<minor>\d+)          # minor version
+        (?P<type>[xXrRsS])?     # version type (optional)
+        (?P<build>\d+)?         # build (optional)
+        """,
+        re.VERBOSE,
+    )
+    re_service_build_respin: re.Pattern[str] = re.compile(
+        r"""
+        (?P<service>[sSdD])?        # service (optional)
+        (?P<service_build>\d+)?     # service build (optional)
+        \.?
+        (?P<service_respin>\d+)?    # service respin (optional)
+        """,
+        re.VERBOSE,
+    )
+    # Set empty params for service pieces and complete them if a second indice exists from the version split
+    # Define isservice, isfrs, isspecial, ismaintenance
+    parsed_version: t.Dict[str, t.Any] = {
+        "isservice": False,
+        "ismaintenance": False,
+        "isfrs": False,
+        "isspecial": False,
+        "service": None,
+        "service_build": None,
+        "service_respin": None,
+    }
+
+    # Juniper junos marks the division between main, minor, type and build from the service build and respin with a -
+    version_core_part, *version_service_part = re.split("-|:", version)
+
+    # Parse out junos into sections that can be used for logic
+    parsed_version.update(re_main_minor_type_build.search(version_core_part).groupdict())  # type:ignore
+
+    if version_service_part:
+        parsed_version.update(re_service_build_respin.search(version_service_part[0]).groupdict())  # type:ignore
+        if parsed_version.get("service", "").lower() == "s":
+            parsed_version["isservice"] = True
+        # Juniper looks at the D in special releases like it's the R in normal releases; Use it as the frs identifier
+        elif parsed_version.get("service").lower() == "d" and (  # type:ignore
+            parsed_version.get("service_build") is None or int(parsed_version.get("service_build", 1)) <= 1
+        ):
+            parsed_version["isfrs"] = True
+
+    if parsed_version.get("type") is None:
+        return parsed_version
+
+    if parsed_version["type"].lower() == "x":
+        parsed_version["isspecial"] = True
+    elif parsed_version["type"].lower() == "s":
+        parsed_version["isservice"] = True
+
+    if parsed_version["type"].lower() == "r" and (
+        parsed_version.get("build") is None or int(parsed_version.get("build")) <= 1  # type:ignore
+    ):
+        parsed_version["isfrs"] = True
+    elif parsed_version["type"].lower() == "r":
+        parsed_version["ismaintenance"] = True
+
```

Comment:
Can we re-evaluate this block? This is alot of conditionals, can we simplify anyway?

Reply:
So, this is actually the re-evaluated block 😂 , it was much larger and very nested.
Juniper has a pretty complex naming system and is why it's the only OS that we are focusing on parsing right now.  Here is a video that kind of goes through the explanation of all the different parts of the version and why there are so many conditionals.

https://www.youtube.com/watch?v=UHW5kiJ9wOg

If you have some suggestions on how we could further improve this block, it would be very welcomed!!